### PR TITLE
Coap integration - coap API client - release-2.0.x

### PR DESCRIPTION
**Pull request by navoda**

### CODE REVIEW

File: components/apimgt-extensions/org.wso2.carbon.apimgt.webapp.publisher/src/main/java/org/wso2/carbon/apimgt/webapp/publisher/APIPublisherStartupHandler.java [hunk 1]
```diff
@@ -55,6 +55,19 @@ public void run() {
                     log.debug("Total number of unpublished APIs: "
                             + APIPublisherDataHolder.getInstance().getUnpublishedApis().size());
                 }
+
+                //coap client bound to the server[the server is in the static default port for now]
+                APIPublisherDataHolder.getInstance().setClient(new CoAPResourceDirectoryClient());
+//                if (APIPublisherDataHolder.getInstance().getClient().isServerConnected()) {
```
Comment (rasika) [inline components/apimgt-extensions/org.wso2.carbon.apimgt.webapp.publisher/src/main/java/org/wso2/carbon/apimgt/webapp/publisher/APIPublisherStartupHandler.java:61]: remove commented lines

File: components/apimgt-extensions/org.wso2.carbon.apimgt.webapp.publisher/src/main/java/org/wso2/carbon/apimgt/webapp/publisher/APIPublisherServiceImpl.java [hunk 2]
```diff
@@ -53,6 +53,8 @@
 
     @Override
     public void publishAPI(final API api) throws APIManagementException, FaultGatewaysException {
+
+        CoAPResourceDirectoryClient client=APIPublisherDataHolder.getInstance().getClient();
```
Comment (rasika) [inline components/apimgt-extensions/org.wso2.carbon.apimgt.webapp.publisher/src/main/java/org/wso2/carbon/apimgt/webapp/publisher/APIPublisherServiceImpl.java:57]: check code formatting of this class

File: components/apimgt-extensions/org.wso2.carbon.apimgt.webapp.publisher/pom.xml [hunk 3]
```diff
@@ -103,9 +103,11 @@
             <groupId>org.wso2.carbon.governance</groupId>
             <artifactId>org.wso2.carbon.governance.lcm</artifactId>
         </dependency>
+        <!-- californium dependancies -->
         <dependency>
-            <groupId>javax.ws.rs</groupId>
-            <artifactId>javax.ws.rs-api</artifactId>
+            <groupId>org.eclipse.californium</groupId>
+            <artifactId>cf-rd</artifactId>
+            <version>1.1.0-SNAPSHOT</version>
```
Comment (rasika) [inline components/apimgt-extensions/org.wso2.carbon.apimgt.webapp.publisher/pom.xml:110]: californium resource directory is not in active development it seems[[1]](https://github.com/eclipse/californium.tools/graphs/contributors). We might need to consider our own implementation of this.

File: components/apimgt-extensions/org.wso2.carbon.apimgt.webapp.publisher/src/main/java/org/wso2/carbon/apimgt/webapp/publisher/CoAPResourceDirectoryClient.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+					payload += ",<" + uriTemplate.getUriTemplate() + ">;rt=\"device\";if=\"" + uriTemplate.getHTTPVerb()
+							+ "\"";
+				}
+
+            }
+
+            Request request = new Request(CoAP.Code.POST, CoAP.Type.CON); //coap request to send the api to server
+
+			request.setURI(this.getURI() + "?ep=" + endpoint + "&d=" + domain+"&con=" +apiURI.getScheme()+"://"+apiURI.getHost()+":"+apiURI.getPort()); //endpoint name, domain and context
+			//request.setURI(this.getURI() + "?ep=" + endpoint + "&d=" + domain); //with coap context
```
Comment (rasika) [inline components/apimgt-extensions/org.wso2.carbon.apimgt.webapp.publisher/src/main/java/org/wso2/carbon/apimgt/webapp/publisher/CoAPResourceDirectoryClient.java:65]: please remove commented line

File: components/apimgt-extensions/org.wso2.carbon.apimgt.webapp.publisher/src/main/java/org/wso2/carbon/apimgt/webapp/publisher/CoAPResourceDirectoryClient.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * This class is used to register the device APIs to the CoAP Resource Directory Server.
+ */
+public class CoAPResourceDirectoryClient extends CoapClient {
+
+	private static final Log log = LogFactory.getLog(CoAPResourceDirectoryClient.class);
+	//TODO - localhost should be ommited and the host must be defined in a config file [next phase - implementing CoAP RD outside the IoT server]
+	private final static String LOCAL_HOST = "localhost";
+    private final static String RESOURCE_DIRECTORY = "rd";
```
Comment (rasika) [inline components/apimgt-extensions/org.wso2.carbon.apimgt.webapp.publisher/src/main/java/org/wso2/carbon/apimgt/webapp/publisher/CoAPResourceDirectoryClient.java:37]: please check the formatting of this class

File: components/apimgt-extensions/org.wso2.carbon.apimgt.webapp.publisher/src/main/java/org/wso2/carbon/apimgt/webapp/publisher/CoAPResourceDirectoryClient.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                log.debug("CoAP server not connected");
+            }
+        }
+    }
+
+
+
+    public boolean isServerConnected() {
+        return true;
+        //return this.ping();
```
Comment (rasika) [inline components/apimgt-extensions/org.wso2.carbon.apimgt.webapp.publisher/src/main/java/org/wso2/carbon/apimgt/webapp/publisher/CoAPResourceDirectoryClient.java:94]: remove commented line

File: pom.xml [hunk 7]
```diff
@@ -1887,6 +1887,11 @@
                 <enabled>false</enabled>
             </releases>
         </repository>
+        <repository>
+            <id>eclipse_snapshots</id>
```
Comment (rasika) [inline pom.xml:1891]: Depending on external third party repositories is not recommended.

**Comment by charithag** [comment]
Close this as feature is backlogged

**Pull request closed without merging**